code/datums/achievements/_achievement_data.dm

```diff
@@ -66,6 +66,8 @@
 	var/datum/award/A = SSachievements.awards[achievement_type]
 	get_data(achievement_type) //Get the current status first if necessary
 	if(istype(A, /datum/award/achievement))
+		if(data[achievement_type]) //You already unlocked it so don't bother running the unlock proc
```

Comment:
I think it would be better for this to pass in old value to on_unlock then handling it in there, then for example scores could have messages on reaching some thresholds or other such things. But this is optional improvement, this is fine as is.